File: cpp/algorithms/graph/mst/kruskal.cpp
#include <alg_std.hpp>
#include <cassert>
using namespace std;

struct UF {
    vector<int> p, r;
    explicit UF(int n) : p(n), r(n, 0) { iota(p.begin(), p.end(), 0); }
    int find(int x) { return p[x] == x ? x : p[x] = find(p[x]); }
    bool unite(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return false;
        if (r[ra] < r[rb]) swap(ra, rb);
        p[rb] = ra;
        if (r[ra] == r[rb]) ++r[ra];
        return true;
    }
};

int kruskal(int n, vector<array<int, 3>> edges) {
    if (n <= 1) return 0;
    sort(edges.begin(), edges.end(), [](auto& a, auto& b) { return a[2] < b[2]; });
    UF uf(n);
    int tot = 0, cnt = 0;
    for (auto [u, v, w] : edges) {
        if (uf.unite(u, v)) {
            tot += w;
            if (++cnt == n - 1) break;
        }
    }
    if (cnt < n - 1) throw runtime_error("graph is not connected");
    return tot;
}

int main() {
    vector<array<int, 3>> e{{0, 1, 4}, {1, 2, 3}, {0, 2, 2}};
    assert(kruskal(3, e) == 5);
    assert(kruskal(1, {}) == 0);
    try {
        kruskal(2, {});
        assert(false);
    } catch (const runtime_error&) {
    }
    cout << "mst OK" << endl;
    return 0;
}

File: python/algorithms/graph/mst/kruskal.py
"""Kruskal 最小生成树。"""

from __future__ import annotations


class UF:
    def __init__(self, n: int) -> None:
        self.p = list(range(n))
        self.r = [0] * n

    def find(self, x: int) -> int:
        while self.p[x] != x:
            self.p[x] = self.p[self.p[x]]
            x = self.p[x]
        return x

    def unite(self, a: int, b: int) -> bool:
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return False
        if self.r[ra] < self.r[rb]:
            ra, rb = rb, ra
        self.p[rb] = ra
        if self.r[ra] == self.r[rb]:
            self.r[ra] += 1
        return True


def kruskal(n: int, edges: list[tuple[int, int, int]]) -> int:
    """edges: (u,v,w)，无向；返回 MST 边权和。要求图连通。"""
    if n <= 1:
        return 0
    edges = sorted(edges, key=lambda t: t[2])
    uf = UF(n)
    total = 0
    cnt = 0
    for u, v, w in edges:
        if uf.unite(u, v):
            total += w
            cnt += 1
            if cnt == n - 1:
                break
    if cnt < n - 1:
        raise ValueError("graph is not connected")
    return total


if __name__ == "__main__":
    # 三角形 0-1:4, 1-2:3, 0-2:2 -> MST 2+3=5
    e = [(0, 1, 4), (1, 2, 3), (0, 2, 2)]
    assert kruskal(3, e) == 5
    assert kruskal(1, []) == 0
    try:
        kruskal(2, [])
        raise AssertionError("expected ValueError")
    except ValueError:
        pass
    print("mst OK")
